Sources/ComposableArchitecture/Observation/ObservableState.swift

```diff
@@ -9,31 +9,81 @@ import Perception
 /// observation support to a type.
 public protocol ObservableState: Perceptible {
   var _$id: ObservableStateID { get }
+  mutating func _$willModify()
 }
 
 /// A unique identifier for a observed value.
 public struct ObservableStateID: Equatable, Hashable, Sendable {
-  private let uuid: UUID
-  private var tag: Int?
+  @usableFromInline
+  var location: UUID {
+    get { self.storage.location.value }
+    set {
+      if isKnownUniquelyReferenced(&self.storage) {
+        self.storage.location.setValue(newValue)
+      } else {
+        self.storage = Storage(location: newValue, tag: self.tag)
+      }
+    }
+  }
+
+  @usableFromInline
+  var tag: Int? {
+    self.storage.tag
+  }
+
+  private var storage: Storage
+
+  @usableFromInline
+  init(location: UUID, tag: Int? = nil) {
+    self.storage = Storage(location: location, tag: tag)
+  }
 
   public init() {
-    self.uuid = UUID()
+    self.init(location: UUID())
   }
 
-  public static let _$inert = Self()
+  public static func == (lhs: Self, rhs: Self) -> Bool {
+    lhs.storage === rhs.storage
+      || lhs.storage.location.value == rhs.storage.location.value
+      && lhs.storage.tag == rhs.storage.tag
+  }
 
-  public func _$tag(_ tag: Int?) -> Self {
-    var copy = self
-    copy.tag = tag
-    return copy
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(self.location)
+    hasher.combine(self.tag)
   }
 
+  @inlinable
   public static func _$id<T>(for value: T) -> Self {
     (value as? any ObservableState)?._$id ?? ._$inert
   }
+
+  @inlinable
   public static func _$id(for value: some ObservableState) -> Self {
     value._$id
   }
+
+  public static let _$inert = Self()
+
+  @inlinable
+  public func _$tag(_ tag: Int?) -> Self {
+    Self(location: self.location, tag: tag)
+  }
+
+  @inlinable
+  public mutating func _$willModify() {
+    self.location = UUID()
+  }
+
+  private final class Storage: Sendable {
+    fileprivate let location: LockIsolated<UUID>
+    fileprivate let tag: Int?
+
+    init(location: UUID = UUID(), tag: Int? = nil) {
+      self.location = LockIsolated(location)
```

Comment:
I might be missing something but if Storage has true CoW semantics it should be unnecessary to lock-isolate it (given that it'll be mutated in place iff you already have exclusive access.)

Reply:
Yeah I think you're right. Pushed a commit here https://github.com/pointfreeco/swift-composable-architecture/pull/2650/commits/f10c8d445aa8c7db73d1606fddf9cc6b1b28e14a.